### DISCUSSION TITLE
[Feature] Show full frame equivalent focal length instead of actual focal length

### DISCUSSION THREAD

**Discussion by jaisinghdhaliwal** [immich-app/immich#10792] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

A photo captured from a phone at 1x zoom would display as 6.8mm, in the iOS photos app it would display as 24mm. I'm not exactly sure what anyone would do with seeing the real focal length when most people tend to compare using full frame equivalents. Another example, a photo taken from my Sony APS-C camera displays as 100mm when the full frame equivalent would show as 150mm. Always displaying full frame equivalent focal length would allow easier comparisons between cameras. 

In exiftool I can see that the field 'Focal Length' displays the real focal length of the cameras (6.8mm and 100mm). There is also a field called 'Focal Length In 35mm Format' which displays the full frame equivalent focal length for both cameras. 

### Platform

- [ ] Server
- [X] Web
- [X] Mobile

**Comment by AngelaDMerkel** [comment]
I would only really like this feature included in mobile phones, as images being taken on more professional cameras are usually taken by people who understand their various relationships to real focal length. I don't want images taken on a medium format camera to lie to me about the focal length, because I know how the FoV changes with the sensor size. On the other hand I don't know all the minuscule size differences between the various lenses on my mobile devices.

**Reply by jaisinghdhaliwal** [reply]
Yes, on iPhones, the zoom values in the camera are displayed as full frame equivalents as well as in the photos app. It’s just in immich it’ll tell you a different focal length. I feel like 350mm would be the best way to standardise the focal length for all cameras including mobile. Professionals tend to use different lenses between APS-C cameras and full frame cameras due to the crop factor, people usually would get a wider lens on the APS-C if they wanted to match the full frame focal length. 

I feel like the focal length shown in immich should also reflect the 1.5x crop in the focal length value as it would match up with other cameras if they all displayed using full frame equivalents. 

I feel like it also makes sense to apply it to medium format cameras too as it would make photos with the same field of view all have the same focal length, regardless of sensor size.

**Reply by SamSamFR** [reply]
+1

But if immich starts displaying "full-frame equivalent" numbers, I suggest the f-ratio should be displayed accordingly. 
Aperture gives the most useful information about bokeh when given as the *actual* aperture (→ mm). 
But converting focal length to FF equivalent while keeping original f-ratio is just wrong.

For instance DXO gives specs : « 24mm equivalent f/1.85-aperture ». What can you make of that f-ratio ? Absolutely nothing without knowing the real f !
The real focal-length is 6.81 mm, so that aperture is actually 3.681 mm, which is 24mm/6.52.
So the bokeh and overall photo characteristics are best expressed as : 
« 24 mm, f/6.52 equivalent ».

tl;dr : either keep displaying the true focal length, or apply the same conversion to the f-ratio displayed.

**Reply by AngelaDMerkel** [reply]
@SamSamFR I disagree with your interpretation in the context of Smart Phones. The field of view can be estimated from the equivalent focal length, but the sensors are so small that the depth of field is still almost irrelevant in the context of photographic decision making. Know how much light was gathered by the lens with an absolute aperture value is more useful

**Reply by SamSamFR** [reply]
> Know how much light was gathered by the lens with an absolute aperture value is more useful

OK, one might not care about DOF for smartphone cameras, I picked bokeh but could have picked anything. At a given field of view, the DOF, total amount of light gathered, background blur … are all a direct, physical, result of the "absolute aperture" as you said (which is in mm, we do agree right ?).

So in the end, you must agree that the f-ratio displayed in "24mm equivalent f/1.85-aperture" would be useless, and immich should either
1. keep true focal length (and f-ratio), even if it not easy to make use of
2. apply multiplier to both focal length **and f-ratio** number → "24 mm, f/6.52 equiv."
3. do whatever with the focal length, and display aboslute aperture, e.g. "24 mm equiv., ⌀3.68 mm"

or any other mix of data, with added precision of the applied multiplier so it is possible for anyone to compute whatever they want from the displayed info,  e.g. : "24 mm, f/6.52 equiv. (3.524×)"

**Reply by AngelaDMerkel** [reply]
> At a given field of view, the DOF, total amount of light gathered, background blur

Except for light gathered, per unit of area on the sensor, those are all features of the lens and sensor in combination; not merely the lens. 

> So in the end, you must agree that the f-ratio displayed in "24mm equivalent f/1.85-aperture" would be useless

I very much disagree. The equivalent focal length provides me with an approximate idea of the field of view and the true aperture with an accurate idea of the amount of light. I am ok with not representing the data in a way which is totally correct because my goal is to represent things to users in a way that is useful. f/6.52 equiv is only equivalent in the context of depth of field and does not represent it's light gathering, and I consider this particular misrepresentation the lesser of the two because depth of field is not a consideration in the context of a smartphone. 

> anyone to compute whatever

I don't think anyone wants to compute anything. I sure don't. In actual cameras I would agree with your proposal, but exif already has a composite spec for "equivalent in 35mm" whereas there is no equivalent field for aperture that I am aware of. It's already settled in the broader debate. In cameras where I know the ratio relative to 35mm I can estimate depth of field easily and in smartphones I would rather know actual light gathering

**Reply by SamSamFR** [reply]
I only have a minute and will write my complete reply later, but in the meantime can you just give one important precision :

> > So in the end, you must agree that the f-ratio displayed in "24mm equivalent f/1.85-aperture" would be useless
> 
> I very much disagree. The equivalent focal length provides me with an approximate idea of the field of view and the true aperture with an accurate idea of the amount of light.

I guess you call "f/1.85" the "true aperture". So please do tell exactly **how** you get an accurate idea of the amount of light, from that number.

_Spoiler : For me, the only true aperture is the size of the hole through which the light goes into the lens, as seen from the outside, and **is expressed in mm**.  And **that** number, together with the field of view, is the only one that actually matters for actual, physical light._

**Reply by AngelaDMerkel** [reply]
> I guess you call "f/1.85" the "true aperture" 

F Stops are already a ratio and provide a close approximation of the amount of light entering in stops, hence the use of this arbitrary and derived unit. T stops provide information about actual light transmittance, but are not common in use for stills photography. The F Stop expresses the aperture and focal length in it's number, which are the two numbers used to derive it. Representing the F Stop in it's depth of field equivalent hides the information related to light gathering from the user and provides very little in the context of depth of field. 

> So please do tell exactly how you get an accurate idea of the amount of light

In case it is unclear, the F Stop and T Stop are very closed to one another, with the T Stop actually measuring transmittance of light. I am happy to use F Stop as a close enough approximation. Providing the 35mm equivalent obscures the amount of light transmitted and only thereafter describes depth of field. 

Websites reporting on the iPhone, for example, describe it using the focal length equivalent in 35mm coupled with the actual aperture (rather than an equivalent). [DxO Mark](https://www.dxomark.com/apple-iphone-16-pro-max-camera-test/) also follows the convention I am suggesting. [As does DP Review.](https://www.dpreview.com/articles/8615681274/iphone-15-pro-vs-iphone-15-camera-test) 

Users are used to having only the focal length equivalised and never the aperture. I am far from the first person to think it makes sense. The exif spec even has a field for my suggestion because it's how this data is always represented. 

For an example: 

An iPhone 16 has an f1.78 lens and 24mm equivalent (as described in DP review). We know that f1.78 is quite bright and provides a lot of light, which is a factor in what makes it better than the f2.2 ultrawide. Choosing to use the aperture equivalent in 35mm is only equivalent in the depth of field, and no longer in the amount of light entering. The same is not true of home 35mm equivalent represents field of view. We all know what a 24mm lens is like, but no clue what a 3mm lens looks like.

**Reply by SamSamFR** [reply]
> > So please do tell exactly how you get an accurate idea of the amount of light
> 
> In case it is unclear, the F Stop and T Stop are very closed to one another

Right, we don't need to discuss transmittance effects on light gathering (and in that case, it would just be multiplying either value by  another factor).

But, see, if I use a 35 mm camera to take a picture in a well lit room, at F/4 and around 1/100s, I know the shot noise of the image, displayed on a 24 inch screen, will be quite ok. 

If I use a μ4/3 camera (2x crop factor) with EQUIVALENT focal-length and f-ratio, still at 1/100s : I'll get the same field of view, same DOF, same background blur  and **same amount of light gathered** (over whole image), thus **same shot noise** and same quality on my same 24" display.

If I used a smartphone camera with the same EQUIVALENT settings, I'd still get equivalent result in the end (i.e. : displayed at same size). [ My smartphone has a 3.52x crop factor sensor, so "F/4 equivalent" means a f/1.1 lens, which it doesn't have, I know. ]

This equivalency works both sides : my 6.81 mm, f/1.85 tiny-sensor lens is 35mm-equivalent to a 24mm, f/6.51 full-frame lens. I know that setup gathers enough light to still be ok for well lit room, but it requires reduced shutter speed or there will be noticeable shot-noise. Although smartphones do all kinds of tricks to counter it (stack exposures, AI noise reduction...), if my subject is moving, quality won't be perfect.

On the other hand, if I only know "24 mm equivalent, f/1.85", then I can't have any such sense. And no-one can without knowing the crop-factor, as that description on its own lacks information to know just how small of a hole light is being captured through. That hole would be 13mm wide on a 35 mm camera, but would be  3 mm on my smartphone - with dramatically less light gathered.


You're right that the (raw) f-number of the lens translates to the amount of light gathering **per sensor area unit** .  But I look at the image when it is displayed on screen, not when it lands on the small sensor area : the total amount of light, spread over my screen, is what conditions the final image quality, not the (temporary) density of light on the sensor area.

You keep implying the 35 mm equivalency only applies to field of view and DOF : it does not. It very much applies to amount of light gathered over the whole sensor, and thus also  **per area of the final image, scaled to your screen or print size**, which is what matters most.


> Websites reporting on the iPhone, for example, describe it using the focal length equivalent in 35mm coupled with the actual aperture (rather than an equivalent). [DxO Mark](https://www.dxomark.com/apple-iphone-16-pro-max-camera-test/) also follows the convention I am suggesting. [As does DP Review.](https://www.dpreview.com/articles/8615681274/iphone-15-pro-vs-iphone-15-camera-test)
> 
> Users are used to having only the focal length equivalised and never the aperture. I am far from the first person to think it makes sense. The exif spec even has a field for my suggestion because it's how this data is always represented.

Usage is indeed to apply the equivalency to the focal length (→ multiply by the crop factor), which makes the hardware specs sounds more impressive, and not to the f-number, where it would make it look much more MEH. 

I'm guessing the companies that sell the hardware very much prefer it that way, and I'm not sure there's any deep truth to be learnt from it except marketing (almost) always win, and habits die hard …

On a sidenote, I noticed [dpreview's iPhone 16 pro review](https://www.dpreview.com/news/8163093616/iphone-16-pro-worth-upgrading-13-14-15-apple-cameras) now displays the crop factor next to the f-number, so I was able to make use of it later in this post. That's  a big change compared to the iPhone 15 review you linked, and could be a sign they realized the raw f-numbers are misleading without it.

> For an example:
> 
> An iPhone 16 has an f1.78 lens and 24mm equivalent (as described in DP review). We know that f1.78 is quite bright

Well, it's bright for its size, but it does not gather much light. A tiny-sensor  f/1.78 lens is probably a well designed lens, but captures less light than, e.g., any APS-C f/4.
You can't shoot fast in low light with a f/1.78 lens designed for a tiny sensor.

> and provides a lot of light, which is a factor in what makes it better than the f2.2 ultrawide. 

Those f-numbers allow you to compare two lenses in the same camera system. But .... you could also make comparison based on equivalent f-numbers. I'm reading the crop factor for the normal is 3.5, and 6.2 for the UW. So equivalent specs would read :  
lens 1 :  24 mm, f/6.23  (35mm equivalent)
lens 2 :  13 mm, f/13.64  (35mm equivalent)

As a matter of fact, it turns out the sensor sizes are different, those were not the same camera system and comparing on f-number alone is very misleading. In fact, **the ultra wide camera of this iPhone is more than two full stops slower than its normal camera**. 

> Choosing to use the aperture equivalent in 35mm is only equivalent in the depth of field, and no longer in the amount of light entering. 

No. The equivalency is very useful to compare the amount of light gathering capacity too. As your example above just demonstrated.

Anyway, this discussion we're having, and the difficulty to reach agreement is nothing new, and has been happening on forums (like dpreview) for at least two decades. 

The equivalency states "the way you can take similar looking photos with two different cameras" : 
Equivalent Focal Length = (Focal Length) x (Crop Factor)
Equivalent F-Number = (F-Number) x (Crop Factor)
Equivalent ISO = (ISO) x (Crop Factor^2)

Some people find this equivalency useful, others don't because it clashes with how they've become used to think about those things.

So I guess the conclusion here is : the best solution in immich would be to let each user decide what informations to display, by a customisable format-string.

**Reply by AngelaDMerkel** [reply]
> thus same shot noise and same quality on my same 24" display.

This makes too many assumptions to be correct. It assumes that you have a bayer filter on all the sensors. It assumes your final images does not apply a crop which affects its apparent visible noise on your tiny display, and it assumes that your are using a digital sensor. I and others don't make images with only digital sensors and in my case I am working with a very wide variety of film stocks in a wide variety of cameras and they don't have shot noise. Any grain which may or may not be represented in the negative is a more or less inherent quality of the development.

A panoramic image taken on a longer lens has a much smaller total diagonal than a 67 negative and would therefore affect whatever equivalency conversion you want to see, but because of how the negative is displayed that grain would be identically visible on your 24" monitor owing to how aspects ratios fit within display boundaries.

> In fact, the ultra wide camera of this iPhone is more than two full stops slower than its normal camera.

The lens is not slower at all. It merely has a smaller sensor behind it, resulting in less total light hitting it, but the lens is no more or less slow. On ILC cameras or on film this is a very important distinction, in my opinion.

> the best solution in immich would be to let each user decide what informations to display, by a customisable format-string.

I am in agreement, because although I understand your argument, I don't think it's an appropriate way of considering all imaging systems and platforms. All this said, I doubt that the immich team will waste their time reading all of our discussion.

**Reply by SamSamFR** [reply]
> > thus same shot noise and same quality on my same 24" display.
> 
> This makes too many assumptions to be correct. 

The whole equivalence kinda relies on ignoring what happens after the light goes thru the aperture. One can consider it's a perfect conversion to some temporary recording, which is then print or displayed to form the final image. Or one can consider it' imperfect, with some degree of differences between devices.

The point of the equivalence is to take out the (big) factors that can be made equivalent from this complicated equations. 
So, sure, if you think that the variations between sensor efficiencies (or between DSLR sensor and silver film, etc.…) make the difference in the optical system a moot point, then the equivalence has no value to you.

> a wide variety of cameras and they don't have shot noise. Any grain which may or may not be represented in the negative is a more or less inherent quality of the development.

We're going on a tangent now, but the shot noise I'm referring to is actually fascinating, and it is very much **in the light** itself, you can't take a picture without it being there. On digital, on film, in your eyes... As I said the whole equivalence only cares about the photons, coming from the outside, through the aperture (and assuming they are then all converted to whatever system you use to record them). The shot noise is fundamentally a quantum fact, that the number of photons will always vary (from the very moment they are emitted by a source), with deviation proportional to the square root of the light intensity. That's the shot noise, and it causes low SNR pictures in low light even if you were using a perfect sensor, or perfect film, or perfect eye cells : SNR=N/√N simply **is** low when the number of photons is low.  

To quote [wikipedia](https://en.wikipedia.org/wiki/Shot_noise#Examples) : 

> Shot noise exists because phenomena such as light and electric current consist of the movement of discrete (also called "quantized") 'packets'. Consider light—a stream of discrete photons—coming out of a laser pointer and hitting a wall to create a visible spot. The fundamental physical processes that govern light emission are such that these photons are emitted from the laser at random times; but the many billions of photons needed to create a spot are so many that the brightness, the number of photons per unit of time, varies only infinitesimally with time. However, if the laser brightness is reduced until only a handful of photons hit the wall every second, the relative fluctuations in number of photons, i.e., brightness, will be significant, just as when tossing a coin a few times. These fluctuations are shot noise.

sensor / film / eye can only add to the photons shot noise. In low light situations, modern cameras add very little, the noise is really in the light itself. When I learnt about that I found it mind-blowing.

> > In fact, the ultra wide camera of this iPhone is more than two full stops slower than its normal camera.
> 
> The lens is not slower at all.

ok, but I said the "ultra wide **camera** of this iPhone" is slower than its other camera. Isolating that lens out of its system, even though it's a fixed focal, with fixed aperture, glued to a small sensor, only brings confusion : if you *could* take it out and place it on a 35 mm sensor, it would still capture not many photons, and send them all to tiny spot in the center. It just will never capture much light. But you can call it fast, in respect of the fact a f/2.2 lens requires a careful design, even if it only uses a few grams of glass.

> > the best solution in immich would be to let each user decide what informations to display, by a customisable format-string.
> 
> I am in agreement, because although I understand your argument, I don't think it's an appropriate way of considering all imaging systems and platforms. All this said, I doubt that the immich team will waste their time reading all of our discussion.

Yeah, but replacing a text field with a format string is an easy modification. Adding a configuration item in the UI is a bit more work. I might just do that when I have a day of free time in a few months.
